src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -37,12 +37,12 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
 
             // determine if we should suppress this run and filter out all the additional files and references if so
             var isGeneratorSuppressed = analyzerConfigOptions.CheckGlobalFlagSet("SuppressRazorSourceGenerator");
-            var additionalTexts = context.AdditionalTextsProvider.EmptyWhen(isGeneratorSuppressed, true);
-            var metadataRefs = context.MetadataReferencesProvider.EmptyWhen(isGeneratorSuppressed, true);
+            var additionalTexts = context.AdditionalTextsProvider.EmptyOrCachedWhen(isGeneratorSuppressed, true);
+            var metadataRefs = context.MetadataReferencesProvider.EmptyOrCachedWhen(isGeneratorSuppressed, true);
 
             var razorSourceGeneratorOptions = analyzerConfigOptions
                 .Combine(parseOptions)
-                .Combine(isGeneratorSuppressed)
+                .SuppressIfNeeded(isGeneratorSuppressed)
```

Comment:
This scenario is handled (we test is here: https://github.com/dotnet/razor/pull/9717/files#diff-21bee42801ceadff30fba484ddb46162fb1af8d819b27b06e2c1ebcd612f4feeR2619)

Some of the inputs all have `EmptyOrCachedWhen` on them: that means when the suppression flag changes, we'll re-evaluate them, and realize that one of the items is different. 

The other inputs are all used in conjunction with `SuppressIfNeeded` that ensures they will also be re-evaluated when that value changes.

Reply:
I amended the test to cover the compilation changing too.

Reply:
The test adds a document, does that change the inputs here (`analyzerConfigOptions` or `parseOptions`)? Because I'm talking specifically about `SuppressIfNeeded` which seems it won't re-run if the inputs change while the suppression was on - since the lambda comparer inside is returning true while suppressed, then when unsuppressed, it compares the old with the new - but the old could have changed during the suppression, hence the old would be equal to new, but could be different from the state when the suppression was off previously. For example:

1. Suppression = off.
1. parseOptions = A -> razorSourceGeneratorOptions = A
1. parseOptions = B -> razorSourceGeneratorOptions = B
1. Suppression = on.
1. parseOptions = A -> razorSourceGeneratorOptions = B (since `SuppressIfNeeded` reports "up to date" due to suppression)
1. Suppression = off -> razorSourceGeneratorOptions = B (since `SuppressIfNeeded` compares the previous value of parseOptions = A with the current value A hence "up to date") - seems wrong, should be changed to A.

Isn't that what could happen?

Reply:
No, that should still work. Because we combine with the configOptions (which has to change to change suppression) `ComputeRazorSourceGeneratorOptions` still runs as needed and updates the options.

Reply:
Added more to the test to cover that scenario.

Reply:
Ok, thanks. Although it sounds like the SuppressIfNeeded works only due to the pipeline being constructed in a specific way (eg, here it works only because one of the inputs contains the suppression check). Might be worthwhile to add a warning to the doc comment of SuppressIfNeeded to make sure it's not used incorrectly in the future.